### PR TITLE
[data] Rename `fully_executed` to `cache` in preparation for streaming by default

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

With streaming execution by default, the semantics of fully_executed() change to become more of a cache API. Rename this accordingly. This also solves the user question of "how do I cache a dataset in memory in streaming mode?"

### CODE REVIEW

File: python/ray/data/dataset.py [hunk 1]
```diff
@@ -3930,7 +3930,32 @@ def __iter__(self):
         return pipe
 
     def fully_executed(self) -> "Dataset[T]":
-        """Force full evaluation of the blocks of this dataset.
+        """Deprecated: use `ds.cache()` instead"""
```
Comment (c21) [inline python/ray/data/dataset.py:3933]: nit: shall we be more explicit with changing `ds.cache()` to `Dataset.cache()` or `cache()`? Technically there's no such a thing of `ds`.

File: release/air_examples/gptj_deepspeed_finetuning/test_myst_doc.py [hunk 2]
```diff
@@ -1 +0,0 @@
-../../../doc/test_myst_doc.py
```
Comment (c21) [inline release/air_examples/gptj_deepspeed_finetuning/test_myst_doc.py:1]: why this change is included?

File: python/ray/train/examples/horovod/horovod_cifar_pbt_example.py [hunk 3]
```diff
@@ -1 +0,0 @@
-../../../../../release/air_tests/horovod/workloads/horovod_tune_test.py
```
Comment (c21) [inline python/ray/train/examples/horovod/horovod_cifar_pbt_example.py:1]: why this change is included here?

**Review by c21 — APPROVED** [review]

File: python/ray/train/tests/pytorch_pbt_failure.py [hunk 4]
```diff
@@ -1 +0,0 @@
-../../../../release/long_running_distributed_tests/workloads/pytorch_pbt_failure.py
```
Comment (c21) [inline python/ray/train/tests/pytorch_pbt_failure.py:1]: why this change is included here?

**Review by richardliaw — APPROVED** [review]

**Review by scv119 — APPROVED** [review]

**Review by Yard1 — APPROVED** [review]

**Review by jianoaix — APPROVED** [review]
